sky/skylet/providers/ibm/node_provider.py

```diff
@@ -0,0 +1,924 @@
+from pathlib import Path
+from uuid import uuid4
+
+from ibm_cloud_sdk_core import ApiException
+from ibm_cloud_sdk_core.authenticators import IAMAuthenticator
+from ibm_vpc import VpcV1
+from typing import Any, Dict, List, Optional
+
+from ray.autoscaler._private.cli_logger import cli_logger
+from ray.autoscaler._private.util import hash_runtime_conf, hash_launch_conf
+from ray.autoscaler.node_provider import NodeProvider
+from ray.autoscaler.tags import (
+    NODE_KIND_HEAD,
+    NODE_KIND_WORKER,
+    TAG_RAY_CLUSTER_NAME,
+    TAG_RAY_NODE_KIND,
+    TAG_RAY_NODE_NAME,
+    TAG_RAY_LAUNCH_CONFIG,
+    TAG_RAY_RUNTIME_CONFIG,
+    TAG_RAY_USER_NODE_TYPE,
+    TAG_RAY_NODE_STATUS,
+    TAG_RAY_FILE_MOUNTS_CONTENTS
+)
+from sky.skylet.providers.ibm.vpc_provider import IBMVPCProvider
+from sky.skylet.providers.ibm.utils import get_logger
+
+logger = get_logger("node_provider_")
+
+INSTANCE_NAME_UUID_LEN = 8
+INSTANCE_NAME_MAX_LEN = 64
+PENDING_TIMEOUT = 3600  # period before a node that's yet to run is deleted.
+VOLUME_TIER_NAME_DEFAULT = "general-purpose"
+# identifies resources created by this package.
+# these resources are deleted alongside the node.
+RAY_RECYCLABLE = "ray-recyclable"
+VPC_TAGS = ".sky-vpc-tags"
+
+
+def _get_vpc_client(endpoint, authenticator):
+    """
+    Creates an IBM VPC python-sdk instance
+    """
+    ibm_vpc_client = VpcV1(version="2022-06-30", authenticator=authenticator)
+    ibm_vpc_client.set_service_url(endpoint + "/v1")
+
+    return ibm_vpc_client
+
+
+def log_in_out(func):
+    """Tracing decorator for debugging purposes"""
+
+    def decorated_func(*args, **kwargs):
+        name = func.__name__
+        logger.debug(
+            f"\n\nEnter {name} from {inspect.stack()[0][3]} "
+            f"{inspect.stack()[1][3]} {inspect.stack()[2][3]} with args: "
+            f"entered with args:\n{pprint(args)} and kwargs {pprint(kwargs)}"
+        )
+        try:
+            result = func(*args, **kwargs)
+            logger.debug(
+                f"Leave {name} from {inspect.stack()[1][3]} with result "
+                f"Func Result:{pprint(result)}\n\n"
+            )
+        except Exception:
+            cli_logger.error(f"Error in {name}")
+            raise
+        return result
+
+    return decorated_func
+
+
+class IBMVPCNodeProvider(NodeProvider):
+    """Node Provider for IBM VPC
+
+    This provider assumes ray-cluster.yaml contains IBM Cloud credentials and
+    all necessary ibm vpc details including existing VPC id, VS image, security
+    group...etc.
+
+    Most convenient way to generate config file is to use `ibm-ray-config` config
+    tool. Install it using `pip install ibm-ray-config`, run it with --pr flag,
+    choose `Ray IBM VPC` and follow interactive wizard.
+
+    Currently, instance tagging is implemented using internal cache
+
+    To communicate with head node from outside cluster private network,
+    `use_hybrid_ips` set to True. Then, floating (external) ip allocated to
```

Comment:
Why is it necessary to have hybrid ips?

I believe SkyPilot will use external ip to communicate with the head node, but will use internal ips for inter-node communication.